configs/named_values.go

```diff
@@ -250,6 +267,152 @@ func (m VariableParsingMode) Parse(name, value string) (cty.Value, hcl.Diagnosti
+	// is required for the condition to return true in a way that should make
+	// sense to a caller of the module.
+	ErrorMessage string
+
+	DeclRange hcl.Range
+}
+
+func decodeVariableValidationBlock(varName string, block *hcl.Block, override bool) (*VariableValidation, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+	vv := &VariableValidation{
+		DeclRange: block.DefRange,
+	}
+
+	if override {
+		// For now we'll just forbid overriding validation blocks, to simplify
+		// the initial design. If we can find a clear use-case for overriding
+		// validations in override files and there's a way to define it that
+		// isn't confusing then we could relax this.
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Can't override variable validation rules",
+			Detail:   "Variable \"validation\" blocks cannot be used in override files.",
+			Subject:  vv.DeclRange.Ptr(),
+		})
+		return vv, diags
+	}
+
+	content, moreDiags := block.Body.Content(variableValidationBlockSchema)
+	diags = append(diags, moreDiags...)
+
+	if attr, exists := content.Attributes["condition"]; exists {
+		vv.Condition = attr.Expr
+
+		// The validation condition can only refer to the variable itself,
+		// to ensure that the variable declaration can't create additional
+		// edges in the dependency graph.
+		goodRefs := 0
+		for _, traversal := range vv.Condition.Variables() {
+			ref, moreDiags := addrs.ParseRef(traversal)
+			if !moreDiags.HasErrors() {
+				if addr, ok := ref.Subject.(addrs.InputVariable); ok {
+					if addr.Name == varName {
+						goodRefs++
+						continue // Reference is valid
+					}
+				}
+			}
+			// If we fall out here then the reference is invalid.
+			diags = diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid reference in variable validation",
+				Detail:   fmt.Sprintf("The condition for variable %q can only refer to the variable itself, using var.%s.", varName, varName),
+				Subject:  traversal.SourceRange().Ptr(),
+			})
+		}
+		if goodRefs < 1 {
+			diags = diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid variable validation condition",
+				Detail:   fmt.Sprintf("The condition for variable %q must refer to var.%s in order to test incoming values.", varName, varName),
+				Subject:  attr.Expr.Range().Ptr(),
+			})
+		}
+	}
+
+	if attr, exists := content.Attributes["error_message"]; exists {
+		moreDiags := gohcl.DecodeExpression(attr.Expr, nil, &vv.ErrorMessage)
+		diags = append(diags, moreDiags...)
+		if !moreDiags.HasErrors() {
+			const errSummary = "Invalid validation error message"
+			switch {
+			case vv.ErrorMessage == "":
+				diags = diags.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  errSummary,
+					Detail:   "An empty string is not a valid nor useful error message.",
+					Subject:  attr.Expr.Range().Ptr(),
+				})
+			case !looksLikeSentences(vv.ErrorMessage):
```

Comment:
I lean towards making this a warning, not an error, from the start. Can you elaborate on your thoughts here?

Reply:
We commonly use warnings in situations where we do expect the user to "fix" it but we're giving them some room to delay fixing it. Since this is a new thing, there is no existing configuration out there relying on being able to use a "not looking like a sentence" message, and so I'm not sure that making it a warning would really help anyone: if you're writing a new configuration you might as well make it warning-free from the outset, right?

I could see us potentially having in future an extra "linting" mechanism that produces messages aimed at the author of a re-usable module but not at _users_ of that module. We don't currently have such a thing, but if we were to add it in future then I'd consider this a good example of something that could get demoted only to a lint error. But we can generally only make things _less_ restrictive over time, so my instinct was to start restrictive here and relax it as we learn more.

Reply:
That makes sense, thank you!